File: LP.Plot.Core/Primitives/Span.cs
namespace LP.Plot.Core.Primitives;

public readonly struct Span
{
    public double Min { get; }
    public double Max { get; }

    public double Length => Max - Min;

    public Span(double min, double max)
    {
        Min = min;
        Max = max;
    }

    public double Mid => (Min + Max) / 2;
    public bool Contains(Span other)
        => Min <= other.Min && other.Max <= Max;

    public override string ToString()
    {
        return $"[Min: {Min}, Max: {Max}, Length: {Length}]";
    }

    public Span ScaleAtCenter(double factor)
    {
        var mid = Mid;
        var newLeftSide = (mid - Min) * factor;
        var newRightSide = (Max - mid) * factor;
        var min = mid - newLeftSide;
        var max = mid + newRightSide;
        return new Span(min, max);
    }
}

File: webgpu/src/primitves/span.ts
export class Span {

    constructor(public min: number, public max: number) { }

    length() {
        return this.max - this.min;
    }

    zoomRelative(factor: number, position: number) {
        var zoomPoint = this.min + this.length() * position;
        this.zoom(factor, zoomPoint);
    }

    zoom(factor: number, position: number) {
        var newLeftSide = (position - this.min) * factor;
        var newRightSide = (this.max - position) * factor;
        this.min = position - newLeftSide;
        this.max = position + newRightSide;
    }

    panRelative(pan: number) {
        let p = pan * this.length();
        this.min += p;
        this.max += p;
    }
}